## Preparing the Data
<span  style="color:purple; font-size:25px">
Loading the Data
</span>

In [140]:
import pandas as pd
import seaborn as sns
import numpy as np
%matplotlib inline

import matplotlib.pyplot as plt

In [141]:
building = pd.read_csv('Datasets/building.csv')
building.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3957 entries, 0 to 3956
Data columns (total 27 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   strap                   3957 non-null   object 
 1   bld_num                 3957 non-null   int64  
 2   effective_year_built    3957 non-null   int64  
 3   design_code             3957 non-null   int64  
 4   design                  3957 non-null   object 
 5   quality                 3957 non-null   object 
 6   quality_code            3957 non-null   int64  
 7   bldg_class_code         3957 non-null   int64  
 8   bldg_class              3957 non-null   object 
 9   construction_type_code  3743 non-null   float64
 10  construction_type       3736 non-null   object 
 11  nbr_bed_room            3957 non-null   float64
 12  nbr_full_baths          3957 non-null   float64
 13  nbr_three_qtr_baths     3957 non-null   float64
 14  nbr_half_baths          3957 non-null   

In [142]:
land = pd.read_csv('Datasets/land.csv')
land.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3433 entries, 0 to 3432
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   strap            3433 non-null   object 
 1   land_class       3433 non-null   int64  
 2   land_class_dscr  3433 non-null   object 
 3   sqft             3433 non-null   int64  
 4   acreage          3433 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 134.2+ KB


In [143]:
property = pd.read_csv('Datasets/property.csv')
property.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3431 entries, 0 to 3430
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   strap           3431 non-null   object 
 1   market_area     3431 non-null   int64  
 2   address         3431 non-null   object 
 3   unincorporated  3431 non-null   bool   
 4   sub_code        3431 non-null   int64  
 5   sub_dscr        3431 non-null   object 
 6   section         3431 non-null   int64  
 7   township        3431 non-null   object 
 8   range           3431 non-null   int64  
 9   mill_levy       3431 non-null   float64
 10  folio           3431 non-null   object 
dtypes: bool(1), float64(1), int64(4), object(5)
memory usage: 271.5+ KB


In [144]:
sales = pd.read_csv('Datasets/sales.csv')
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9598 entries, 0 to 9597
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   strap             9598 non-null   object
 1   transaction_date  9598 non-null   object
 2   sales_cd          9598 non-null   object
 3   sales_cd_dscr     9598 non-null   object
 4   price             9598 non-null   int64 
dtypes: int64(1), object(4)
memory usage: 375.0+ KB


In [145]:
time_trend_adj = pd.read_csv('Datasets/time_trend_adjustments.csv')
time_trend_adj.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       60 non-null     int64  
 1   year             60 non-null     int64  
 2   month            60 non-null     int64  
 3   market_area_101  60 non-null     float64
 4   market_area_102  60 non-null     float64
 5   market_area_103  60 non-null     float64
 6   market_area_104  60 non-null     float64
 7   market_area_105  60 non-null     float64
 8   market_area_106  60 non-null     float64
 9   market_area_107  60 non-null     float64
 10  market_area_108  60 non-null     float64
 11  market_area_109  60 non-null     float64
dtypes: float64(9), int64(3)
memory usage: 5.8 KB


In [146]:
valuations = pd.read_csv('Datasets/valuations.csv')
valuations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3431 entries, 0 to 3430
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   strap                3431 non-null   object 
 1   tax_yr               3431 non-null   int64  
 2   bld_appraised_val    3431 non-null   int64  
 3   land_appraised_val   3431 non-null   int64  
 4   total_appraised_val  3431 non-null   int64  
 5   bld_assessed_val     3429 non-null   float64
 6   land_assessed_val    0 non-null      float64
 7   total_assessed_val   3431 non-null   int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 214.6+ KB


<span  style="color:purple; font-size:25px">
Tidying the Data
</span>

#### Datetimes

In [147]:
# Note that the dates in `time_trend_adj` and `sales` are not datetime data types.  Let's correct that using `pd.to_datetime`
# and assign missing day values to 1.

time_trend_adj['date'] = pd.to_datetime(time_trend_adj[['year', 'month']].assign(day = 1))

In [148]:
# Now remove the redundant columns.
time_trend_adj.drop(columns = ['year', 'month', 'Unnamed: 0'], inplace = True)
time_trend_adj.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   market_area_101  60 non-null     float64       
 1   market_area_102  60 non-null     float64       
 2   market_area_103  60 non-null     float64       
 3   market_area_104  60 non-null     float64       
 4   market_area_105  60 non-null     float64       
 5   market_area_106  60 non-null     float64       
 6   market_area_107  60 non-null     float64       
 7   market_area_108  60 non-null     float64       
 8   market_area_109  60 non-null     float64       
 9   date             60 non-null     datetime64[ns]
dtypes: datetime64[ns](1), float64(9)
memory usage: 4.8 KB


In [149]:
# Now do the same for the `sales` data frame.
sales['date'] = pd.to_datetime(sales['transaction_date'])
sales.drop(columns = 'transaction_date', inplace = True)
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9598 entries, 0 to 9597
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   strap          9598 non-null   object        
 1   sales_cd       9598 non-null   object        
 2   sales_cd_dscr  9598 non-null   object        
 3   price          9598 non-null   int64         
 4   date           9598 non-null   datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 375.0+ KB


#### Preparing the Market Adjustments

In [150]:
# Let's look at the `time_trend_adj` data
time_trend_adj.head()

market_area_101  market_area_102  market_area_103  market_area_104  \
0           1.4591           1.6070           1.6384           1.5174   
1           1.4485           1.5989           1.6288           1.5075   
2           1.4379           1.5908           1.6193           1.4977   
3           1.4273           1.5828           1.6099           1.4880   
4           1.4169           1.5748           1.6005           1.4783   

   market_area_105  market_area_106  market_area_107  market_area_108  \
0           1.5532           1.5728           1.5938           1.2996   
1           1.5481           1.5654           1.5830           1.2996   
2           1.5430           1.5579           1.5723           1.2996   
3           1.5379           1.5505           1.5616           1.2996   
4           1.5329           1.5431           1.5510           1.2996   

   market_area_109       date  
0           1.4132 2017-07-01  
1           1.4132 2017-08-01  
2           1.4132 2017-09-01  
3           1.3913 2017-10-01  
4           1.3840 2017-11-01

In order to merge this data frame with `sales` to create an adjusted sales price, we need to change this so there are `market_area`
and `multiplier` columns

In [151]:
time_trend_adj_clean = time_trend_adj.melt(id_vars = 'date',
                                           var_name = 'market_area',
                                           value_name = 'multiplier')

time_trend_adj_clean

date      market_area  multiplier
0   2017-07-01  market_area_101      1.4591
1   2017-08-01  market_area_101      1.4485
2   2017-09-01  market_area_101      1.4379
3   2017-10-01  market_area_101      1.4273
4   2017-11-01  market_area_101      1.4169
..         ...              ...         ...
535 2022-02-01  market_area_109      1.0296
536 2022-03-01  market_area_109      1.0147
537 2022-04-01  market_area_109      1.0000
538 2022-05-01  market_area_109      1.0000
539 2022-06-01  market_area_109      1.0000

[540 rows x 3 columns]

In [152]:
# Now use regular expressions to extract the market area number
time_trend_adj_clean['market_area'] = time_trend_adj_clean['market_area'].str.extract(r'(\d+)$').astype(int)
time_trend_adj_clean

date  market_area  multiplier
0   2017-07-01          101      1.4591
1   2017-08-01          101      1.4485
2   2017-09-01          101      1.4379
3   2017-10-01          101      1.4273
4   2017-11-01          101      1.4169
..         ...          ...         ...
535 2022-02-01          109      1.0296
536 2022-03-01          109      1.0147
537 2022-04-01          109      1.0000
538 2022-05-01          109      1.0000
539 2022-06-01          109      1.0000

[540 rows x 3 columns]

#### Adjusting the Sales Prices

In [153]:
# Merge the `sales` and `property` data frames to prepare for a row-wise calculation
sales_w_market_area = sales.merge(property[['strap', 'market_area']],
                                  on = 'strap',
                                  how = 'left')
sales_w_market_area

strap sales_cd sales_cd_dscr    price       date  market_area
0     R0000008        Q     qualified    65000 1978-01-03          102
1     R0000019        U   unqualified    75000 1980-08-13          109
2     R0000019        U   unqualified   110600 1985-04-08          109
3     R0000019        Q     qualified   126400 1992-04-22          109
4     R0000019        Q     qualified   332000 2000-06-02          109
...        ...      ...           ...      ...        ...          ...
9593  R0610553        Q     qualified  2872600 2021-07-06          102
9594  R0610553        Q     qualified  3450000 2023-02-09          102
9595  R0612718        U   unqualified   910000 2021-05-10          109
9596  R0613548        Q     qualified   850000 2020-09-14          102
9597  R0616832        Q     qualified   710000 2023-06-26          101

[9598 rows x 6 columns]

In [154]:
# After attempting to merge this column with `time_trend_adj_clean` using the `date` (as datetime) and `market_area` columns,
# we ran into issues with many new untidy columns being created.

# Therefore, we will create `year` and `month` columns for each data frame and use them to merge our final two data frames

sales_w_market_area['year'] = sales_w_market_area['date'].dt.year
sales_w_market_area['month'] = sales_w_market_area['date'].dt.month
sales_w_market_area

strap sales_cd sales_cd_dscr    price       date  market_area  year  \
0     R0000008        Q     qualified    65000 1978-01-03          102  1978   
1     R0000019        U   unqualified    75000 1980-08-13          109  1980   
2     R0000019        U   unqualified   110600 1985-04-08          109  1985   
3     R0000019        Q     qualified   126400 1992-04-22          109  1992   
4     R0000019        Q     qualified   332000 2000-06-02          109  2000   
...        ...      ...           ...      ...        ...          ...   ...   
9593  R0610553        Q     qualified  2872600 2021-07-06          102  2021   
9594  R0610553        Q     qualified  3450000 2023-02-09          102  2023   
9595  R0612718        U   unqualified   910000 2021-05-10          109  2021   
9596  R0613548        Q     qualified   850000 2020-09-14          102  2020   
9597  R0616832        Q     qualified   710000 2023-06-26          101  2023   

      month  
0         1  
1         8  
2         4  
3         4  
4         6  
...     ...  
9593      7  
9594      2  
9595      5  
9596      9  
9597      6  

[9598 rows x 8 columns]

In [155]:
# Now do the same for `time_trend_adj_clean`

time_trend_adj_clean['year'] = time_trend_adj_clean['date'].dt.year
time_trend_adj_clean['month'] = time_trend_adj_clean['date'].dt.month
time_trend_adj_clean

date  market_area  multiplier  year  month
0   2017-07-01          101      1.4591  2017      7
1   2017-08-01          101      1.4485  2017      8
2   2017-09-01          101      1.4379  2017      9
3   2017-10-01          101      1.4273  2017     10
4   2017-11-01          101      1.4169  2017     11
..         ...          ...         ...   ...    ...
535 2022-02-01          109      1.0296  2022      2
536 2022-03-01          109      1.0147  2022      3
537 2022-04-01          109      1.0000  2022      4
538 2022-05-01          109      1.0000  2022      5
539 2022-06-01          109      1.0000  2022      6

[540 rows x 5 columns]

In [156]:
time_trend_adj_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 540 entries, 0 to 539
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         540 non-null    datetime64[ns]
 1   market_area  540 non-null    int32         
 2   multiplier   540 non-null    float64       
 3   year         540 non-null    int64         
 4   month        540 non-null    int64         
dtypes: datetime64[ns](1), float64(1), int32(1), int64(2)
memory usage: 19.1 KB


In [157]:
# Now merge this with the `time_trend_adj_clean` data frame to add in the multiplier
sales_w_mult = sales_w_market_area.merge(time_trend_adj_clean[['year', 'month', 'market_area', 'multiplier']],
                                         on = ['year', 'month', 'market_area'],
                                         how = 'left').drop(columns = ['year', 'month'])
sales_w_mult

strap sales_cd sales_cd_dscr    price       date  market_area  \
0     R0000008        Q     qualified    65000 1978-01-03          102   
1     R0000019        U   unqualified    75000 1980-08-13          109   
2     R0000019        U   unqualified   110600 1985-04-08          109   
3     R0000019        Q     qualified   126400 1992-04-22          109   
4     R0000019        Q     qualified   332000 2000-06-02          109   
...        ...      ...           ...      ...        ...          ...   
9593  R0610553        Q     qualified  2872600 2021-07-06          102   
9594  R0610553        Q     qualified  3450000 2023-02-09          102   
9595  R0612718        U   unqualified   910000 2021-05-10          109   
9596  R0613548        Q     qualified   850000 2020-09-14          102   
9597  R0616832        Q     qualified   710000 2023-06-26          101   

      multiplier  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
9593      1.2167  
9594         NaN  
9595      1.1910  
9596      1.3453  
9597         NaN  

[9598 rows x 7 columns]

Now that we have all the necessary columns in one data frame, we can create a column for the adjusted sales price

In [158]:
sales_w_mult['adjusted_sales_price'] = np.where(sales_w_mult['multiplier'].isnull(), 
                                                sales_w_mult['price'],
                                                sales_w_mult['price'] * sales_w_mult['multiplier']).astype(int)
sales_w_mult

strap sales_cd sales_cd_dscr    price       date  market_area  \
0     R0000008        Q     qualified    65000 1978-01-03          102   
1     R0000019        U   unqualified    75000 1980-08-13          109   
2     R0000019        U   unqualified   110600 1985-04-08          109   
3     R0000019        Q     qualified   126400 1992-04-22          109   
4     R0000019        Q     qualified   332000 2000-06-02          109   
...        ...      ...           ...      ...        ...          ...   
9593  R0610553        Q     qualified  2872600 2021-07-06          102   
9594  R0610553        Q     qualified  3450000 2023-02-09          102   
9595  R0612718        U   unqualified   910000 2021-05-10          109   
9596  R0613548        Q     qualified   850000 2020-09-14          102   
9597  R0616832        Q     qualified   710000 2023-06-26          101   

      multiplier  adjusted_sales_price  
0            NaN                 65000  
1            NaN                 75000  
2            NaN                110600  
3            NaN                126400  
4            NaN                332000  
...          ...                   ...  
9593      1.2167               3495092  
9594         NaN               3450000  
9595      1.1910               1083810  
9596      1.3453               1143505  
9597         NaN                710000  

[9598 rows x 8 columns]

In order to conduct time series analysis, we should set the index to the date

In [159]:
sales_adj = sales_w_mult.set_index('date')
sales_adj

strap sales_cd sales_cd_dscr    price  market_area  multiplier  \
date                                                                            
1978-01-03  R0000008        Q     qualified    65000          102         NaN   
1980-08-13  R0000019        U   unqualified    75000          109         NaN   
1985-04-08  R0000019        U   unqualified   110600          109         NaN   
1992-04-22  R0000019        Q     qualified   126400          109         NaN   
2000-06-02  R0000019        Q     qualified   332000          109         NaN   
...              ...      ...           ...      ...          ...         ...   
2021-07-06  R0610553        Q     qualified  2872600          102      1.2167   
2023-02-09  R0610553        Q     qualified  3450000          102         NaN   
2021-05-10  R0612718        U   unqualified   910000          109      1.1910   
2020-09-14  R0613548        Q     qualified   850000          102      1.3453   
2023-06-26  R0616832        Q     qualified   710000          101         NaN   

            adjusted_sales_price  
date                              
1978-01-03                 65000  
1980-08-13                 75000  
1985-04-08                110600  
1992-04-22                126400  
2000-06-02                332000  
...                          ...  
2021-07-06               3495092  
2023-02-09               3450000  
2021-05-10               1083810  
2020-09-14               1143505  
2023-06-26                710000  

[9598 rows x 7 columns]

#### Ensuring the Rest of the Data is Tidy

##### `Valuations`

In [160]:
valuations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3431 entries, 0 to 3430
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   strap                3431 non-null   object 
 1   tax_yr               3431 non-null   int64  
 2   bld_appraised_val    3431 non-null   int64  
 3   land_appraised_val   3431 non-null   int64  
 4   total_appraised_val  3431 non-null   int64  
 5   bld_assessed_val     3429 non-null   float64
 6   land_assessed_val    0 non-null      float64
 7   total_assessed_val   3431 non-null   int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 214.6+ KB


`land_assessed_val` has no non-null values, but we know that `total_assessed_val` is the sum of it and `bld_assessed_val`, so we can fill in the NA values with their proper amounts.  There are two buildings that have a non-null `total_assessed_val`, but a null `bld_assessed_val`.  Since we still have the total, we will still keep these records.

In [161]:
valuations['land_assessed_val'] = valuations['total_assessed_val'] - valuations['bld_assessed_val']
valuations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3431 entries, 0 to 3430
Data columns (total 8 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   strap                3431 non-null   object 
 1   tax_yr               3431 non-null   int64  
 2   bld_appraised_val    3431 non-null   int64  
 3   land_appraised_val   3431 non-null   int64  
 4   total_appraised_val  3431 non-null   int64  
 5   bld_assessed_val     3429 non-null   float64
 6   land_assessed_val    3429 non-null   float64
 7   total_assessed_val   3431 non-null   int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 214.6+ KB


In [162]:
# Our code appears to have worked, but let's look at a few examples to be sure
valuations.sample(5)

strap  tax_yr  bld_appraised_val  land_appraised_val  \
767   R0002599    2024             869300             1187000   
1145  R0003796    2024             118150             1063350   
980   R0003303    2024              86400              873600   
418   R0001319    2024             369000             1005100   
2307  R0007465    2024             584800              752500   

      total_appraised_val  bld_assessed_val  land_assessed_val  \
767               2056300           58243.0            75844.0   
1145              1181500            7916.0            67560.0   
980                960000            5789.0            54846.0   
418               1374100           24723.0            63657.0   
2307              1337300           39182.0            46732.0   

      total_assessed_val  
767               134087  
1145               75476  
980                60635  
418                88380  
2307               85914

##### `Property`

In [163]:
property.sample(10)

strap  market_area                        address  unincorporated  \
2313  R0007471          109      1700 BLUFF ST, BOULDER CO           False   
1928  R0006299          103         903 8TH ST, BOULDER CO           False   
2785  R0008911          105    428 PLEASANT ST, BOULDER CO           False   
1968  R0006416          109  2370 PANORAMA AVE, BOULDER CO           False   
2619  R0008385          102       1067 13TH ST, BOULDER CO           False   
1515  R0004974          105        1320 6TH ST, BOULDER CO           False   
1266  R0004167          103     895 AURORA AVE, BOULDER CO           False   
167   R0000545          103       920 GRANT PL, BOULDER CO           False   
805   R0002718          101       2019 GOSS ST, BOULDER CO           False   
1470  R0004873          103  1074 ROSE HILL DR, BOULDER CO           False   

      sub_code                         sub_dscr  section township  range  \
2313      7981                 SUNSET HILL - BO       30       1N     70   
1928      8675              WEST ROSE HILL - BO       36       1N     71   
2785      8424                     VERMONT - BO       36       1N     71   
1968      6049     PANORAMA HEIGHTS REPLAT - BO       30       1N     70   
2619      8343            UNIVERSITY PLACE - BO       31       1N     70   
1515      8424                     VERMONT - BO       36       1N     71   
1266      6981                   ROSE HILL - BO       36       1N     71   
167       6981                   ROSE HILL - BO       36       1N     71   
805       1913                      CULVER - BO       30       1N     70   
1470      9901  TR, NBR 101-103,105,107,109,170       36       1N     71   

      mill_levy         folio  
2313     86.359  146330226006  
1928     86.359  146136410017  
2785     86.359  146136138005  
1968     86.359  146330120006  
2619     86.359  146331339012  
1515     86.359  146136133005  
1266     86.359  146136414009  
167      86.359  146136414012  
805      86.359  146330817007  
1470     86.359  146136400031

The `property` data set appears to be tidy

##### `Land`

In [164]:
land.sample(10)

strap  land_class       land_class_dscr   sqft  acreage
1012  R0003401        1115    DUP/TRIPLEXES-LAND  12139    0.279
1497  R0004932        1112  SINGLE FAM.RES.-LAND   4541    0.104
2056  R0006696        1112  SINGLE FAM.RES.-LAND  10127    0.232
506   R0001634        1112  SINGLE FAM.RES.-LAND   5840    0.134
2521  R0008121        1112  SINGLE FAM.RES.-LAND   7876    0.181
2085  R0006772        1112  SINGLE FAM.RES.-LAND  12479    0.286
1740  R0005724        1112  SINGLE FAM.RES.-LAND  10864    0.249
663   R0002195        1112  SINGLE FAM.RES.-LAND  10072    0.231
1875  R0006134        1112  SINGLE FAM.RES.-LAND   6367    0.146
1077  R0003580        1112  SINGLE FAM.RES.-LAND   3225    0.074

In [165]:
land['land_class_dscr'].value_counts()

SINGLE FAM.RES.-LAND    2927
DUP/TRIPLEXES-LAND       503
MANUF HOME PARK LAND       2
VACANT RES LOTS            1
Name: land_class_dscr, dtype: int64

The `land` data set appears to be tidy

##### `Building`

In [166]:
building.sample(10)

strap  bld_num  effective_year_built  design_code  \
218   R0000568        1                  1965           10   
1128  R0003245        1                  1975           20   
1648  R0004635        1                  1990           20   
2645  R0007308        2                  2014           60   
1179  R0003369        1                  1975           20   
3328  R0009063        1                  2005           10   
2329  R0006492        1                  1995           20   
3518  R0075737        1                  2005          120   
562   R0001517        1                  2015           20   
3940  R0606836        1                  2016           20   

                      design       quality  quality_code  bldg_class_code  \
218          1 Story - Ranch      AVERAGE             30             1212   
1128               2-3 Story      GOOD ++             42             1212   
1648               2-3 Story         GOOD             40             1212   
2645                  Studio  VERY GOOD +             51             1212   
1179               2-3 Story   AVERAGE ++             32             1212   
3328         1 Story - Ranch       GOOD +             41             1212   
2329               2-3 Story    EXCELLENT             60             1212   
3518  MULTI STORY- TOWNHOUSE    VERY GOOD             50             1212   
562                2-3 Story    EXCELLENT             60             1212   
3940               2-3 Story    VERY GOOD             50             1212   

                       bldg_class  construction_type_code  ... mainfloor_sqft  \
218   SINGLE FAM RES IMPROVEMENTS                   310.0  ...          840.0   
1128  SINGLE FAM RES IMPROVEMENTS                   320.0  ...         1296.0   
1648  SINGLE FAM RES IMPROVEMENTS                   310.0  ...         1482.0   
2645  SINGLE FAM RES IMPROVEMENTS                   310.0  ...            NaN   
1179  SINGLE FAM RES IMPROVEMENTS                   310.0  ...          990.0   
3328  SINGLE FAM RES IMPROVEMENTS                   320.0  ...         1740.0   
2329  SINGLE FAM RES IMPROVEMENTS                   320.0  ...         1603.0   
3518  SINGLE FAM RES IMPROVEMENTS                   310.0  ...          824.0   
562   SINGLE FAM RES IMPROVEMENTS                   310.0  ...         2556.0   
3940  SINGLE FAM RES IMPROVEMENTS                   310.0  ...         1016.0   

      bsmt_sqft  bsmt_type_code                              bsmt_type  \
218       776.0             BSF    SUBTERRANEAN BASEMENT FINISHED AREA   
1128      528.0             BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA   
1648      370.0             BSF    SUBTERRANEAN BASEMENT FINISHED AREA   
2645        NaN             NaN                                    NaN   
1179      529.0             BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA   
3328     1196.0             BWF        WALK-OUT BASEMENT FINISHED AREA   
2329     1981.0             BGF          GARDEN BASEMENT FINISHED AREA   
3518      740.0             BSF    SUBTERRANEAN BASEMENT FINISHED AREA   
562       692.0             BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA   
3940     2208.0             BGF          GARDEN BASEMENT FINISHED AREA   

      car_storage_type_code      car_storage_type  ext_wall_code  \
218                     GRD       DETACHED GARAGE          100.0   
1128                    NaN                   NaN           70.0   
1648                    NaN                   NaN           10.0   
2645                    GRD       DETACHED GARAGE           10.0   
1179                    NaN                   NaN           10.0   
3328                    GRA  ATTACHED GARAGE AREA           30.0   
2329                    GRD       DETACHED GARAGE           70.0   
3518                    GRC          CARPORT AREA           30.0   
562                     GRD       DETACHED GARAGE           10.0   
3940                    GRA  ATTACHED GARAGE AREA           30.0   

                 ext_w

Some entries are capitalized title-style and some are all uppercase. For consistancy and regex simplicity, we will make them all uppercase

In [167]:
building = building.applymap(lambda x: x.upper() if isinstance(x, str) else x)
building.sample(5)

strap  bld_num  effective_year_built  design_code           design  \
3015  R0008276        1                  1975           10  1 STORY - RANCH   
711   R0001986        1                  2000           20        2-3 STORY   
1684  R0004738        1                  1980           20        2-3 STORY   
863   R0002437        1                  1975           20        2-3 STORY   
2005  R0005646        1                  2003           20        2-3 STORY   

         quality  quality_code  bldg_class_code                   bldg_class  \
3015    AVERAGE             30             1212  SINGLE FAM RES IMPROVEMENTS   
711   EXCELLENT             60             1212  SINGLE FAM RES IMPROVEMENTS   
1684       GOOD             40             1212  SINGLE FAM RES IMPROVEMENTS   
863        GOOD             40             1212  SINGLE FAM RES IMPROVEMENTS   
2005    GOOD ++             42             1212  SINGLE FAM RES IMPROVEMENTS   

      construction_type_code  ... mainfloor_sqft  bsmt_sqft  bsmt_type_code  \
3015                   310.0  ...         1448.0        NaN             NaN   
711                    320.0  ...         1027.0      333.0             BSU   
1684                   320.0  ...         1434.0      867.0             BSF   
863                    310.0  ...          984.0        NaN             NaN   
2005                   310.0  ...         1487.0      700.0             BSF   

                                  bsmt_type  car_storage_type_code  \
3015                                    NaN                    GRD   
711   SUBTERRANEAN BASEMENT UNFINISHED AREA                    NaN   
1684    SUBTERRANEAN BASEMENT FINISHED AREA                    GRC   
863                                     NaN                    GRD   
2005    SUBTERRANEAN BASEMENT FINISHED AREA                    GRD   

      car_storage_type  ext_wall_code          ext_wall     ac heating  
3015   DETACHED GARAGE           10.0  FRAME WOOD/SHAKE  False    True  
711                NaN           70.0    BRICK ON BLOCK  False    True  
1684      CARPORT AREA           70.0    BRICK ON BLOCK   True    True  
863    DETACHED GARAGE           10.0  FRAME WOOD/SHAKE  False    True  
2005   DETACHED GARAGE           10.0  FRAME WOOD/SHAKE  False    True  

[5 rows x 27 columns]

#### Adding New Features to the Data

##### Finished Basements

One of the central features that the legislators wanted us to look at was whether or not homes have finished basements, and how i could affect home value.

In [168]:
# Let's examine the different values of basement types that have been recorded
building['bsmt_type'].value_counts()

SUBTERRANEAN BASEMENT FINISHED AREA           965
SUBTERRANEAN BASEMENT UNFINISHED AREA         538
WALK-OUT BASEMENT FINISHED AREA               440
GARDEN BASEMENT FINISHED AREA                 184
LOWER LVL GARDEN FINISHED (BI-SPLIT LVL)       54
WALK-OUT BASEMENT UNFINISHED AREA              31
LOWER LVL WALKOUT FINISHED (BI-SPLIT LVL)      15
GARDEN BASEMENT UNFINISHED AREA                12
LOWER LVL GARDEN UNFINISHED (BI-SPLIT LVL)      2
Name: bsmt_type, dtype: int64

In [169]:
building['bsmt_type_code'].value_counts()

BSF    965
BSU    538
BWF    440
BGF    184
LGF     54
BWU     31
LWF     15
BGU     12
LGU      2
Name: bsmt_type_code, dtype: int64

Since basements are explicitly labeled as "finished" or "unfinished", and their codes end in "F" or "U" accordingly, we can create a boolean column to quickly show whether each building has a finished basement.

In [170]:
building['bsmt_finished'] = np.where(building['bsmt_type_code'].str.contains(r'U$', na = False) | building['bsmt_type_code'].isna(),
                                     False,
                                     True)

In [171]:
# Verify that it this column was properly created by running this code multiple times and verifying
building[['bsmt_type_code', 'bsmt_type', 'bsmt_finished']].sample(10)

bsmt_type_code                              bsmt_type  bsmt_finished
764             NaN                                    NaN          False
3431            NaN                                    NaN          False
729             BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA          False
3492            NaN                                    NaN          False
245             BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA          False
747             BSF    SUBTERRANEAN BASEMENT FINISHED AREA           True
1785            BSU  SUBTERRANEAN BASEMENT UNFINISHED AREA          False
1036            BSF    SUBTERRANEAN BASEMENT FINISHED AREA           True
1704            NaN                                    NaN          False
3388            NaN                                    NaN          False

##### Number of Houses on the Property

Another factor in determining the value of a home is how many buildings are on the property.  We will create a column to record this data.

In [172]:
# How many occurences are there of multiple buildings on a property? What is the proportion?
print(building['bld_num'].value_counts())
print(building['bld_num'].value_counts().iloc[1:4].sum())
building['bld_num'].value_counts().iloc[1:4].sum() / building['bld_num'].value_counts().sum()

1    3429
2     496
3      30
4       2
Name: bld_num, dtype: int64
528


0.133434420015163

Approximately 13.3% (528 properties) of the properties have multiple buildings, enough that to analyze this further.

In [173]:
# Create the `num_buildings` column and populate it with the count of each property (strap)
building['num_buildings'] = building.groupby('strap')['strap'].transform('count')
building.sample(10)
building['num_buildings'].value_counts()

1    2933
2     932
3      84
4       8
Name: num_buildings, dtype: int64

It should be noted that while we normally would not put aggregate values in each occurence of a property, the goal is to join this aggregate data into another data frame to perform analysis.  So, this is a preliminary step to prevent later complications

##### Total Number of Bathrooms

When assessing a home, preliminary data generally does not detail the number of each type of bathroom (half, three-quarter, full).  Instead, the total number is used, so we will create a variable to contain just that by summing the scaled number of bathrooms.

In [174]:
building['num_bath'] = building['nbr_full_baths'] + (0.75 * building['nbr_three_qtr_baths']) + (0.5 * building['nbr_half_baths'])
building[['nbr_full_baths', 'nbr_three_qtr_baths', 'nbr_half_baths', 'num_bath']].sample(10)

nbr_full_baths  nbr_three_qtr_baths  nbr_half_baths  num_bath
2044             1.0                  2.0             0.0      2.50
2811             1.0                  1.0             0.0      1.75
244              0.0                  0.0             0.0      0.00
2693             2.0                  0.0             0.0      2.00
3149             1.0                  1.0             0.0      1.75
1430             0.0                  4.0             1.0      3.50
2323             2.0                  0.0             1.0      2.50
310              2.0                  1.0             0.0      2.75
3375             0.0                  0.0             0.0      0.00
3623             2.0                  1.0             1.0      3.25

For consistency, since we have created `num_buildings` and `num_bath`, we should rename our `nbr_bed_room` column accordingly

In [175]:
building.rename(columns = {'nbr_bed_room': 'num_bed'}, inplace = True)
building[['strap', 'num_bed']].sample(5)

strap  num_bed
1135  R0003258      3.0
1786  R0004971      2.0
2413  R0006721      2.0
1188  R0003401     12.0
3149  R0008598      2.0

#### Filtering and Joining the Data

In [176]:
building['design'].value_counts()

2-3 STORY                                  2139
1 STORY - RANCH                            1183
STUDIO                                      242
MULTI STORY- TOWNHOUSE                      112
TOOL SHED                                    91
SPLIT-LEVEL                                  79
GARAGE DETACHED RESIDENTIAL                  66
EQUIPMENT SHED                               13
BI-LEVEL                                     11
EQUIPMENT (SHOP) BUILDING                     8
GREENHOUSE                                    6
OTHER STRUCTURE                               2
GENERAL PURPOSE BARN                          1
STORAGE SHED (PREFABRICATED)                  1
1-STORY TWNHM                                 1
GRNHS, HOOP, ARCH-RIB, MED(4500-9000SF)       1
MODULAR                                       1
Name: design, dtype: int64

There are many different design types, but for the scope of this analysis we will focus only on *dwellings*.  So let's create a filter.

In [177]:
building_dwellings = building[(building['design'] == '2-3 STORY') |
                              (building['design'] == '1 STORY - RANCH') |
                              (building['design'] == 'STUDIO') |
                              (building['design'] == 'MULTISTORY- TOWNHOUSE') |
                              (building['design'] == 'SPLIT-LEVEL') |
                              (building['design'] == 'BI-LEVEL') |
                              (building['design'] == '1-STORY TWNHM') |
                              (building['design'] == 'MODULAR')]
building_dwellings.sample(5)

strap  bld_num  effective_year_built  design_code           design  \
2839  R0007800        2                  2000           10  1 STORY - RANCH   
3656  R0094084        1                  1960           10  1 STORY - RANCH   
3310  R0009030        1                  1956           10  1 STORY - RANCH   
1824  R0005085        1                  2000           10  1 STORY - RANCH   
2354  R0006559        1                  2018           20        2-3 STORY   

          quality  quality_code  bldg_class_code                   bldg_class  \
2839        GOOD             40             1212  SINGLE FAM RES IMPROVEMENTS   
3656     AVERAGE             30             1212  SINGLE FAM RES IMPROVEMENTS   
3310        GOOD             40             1212  SINGLE FAM RES IMPROVEMENTS   
1824  AVERAGE ++             32             1212  SINGLE FAM RES IMPROVEMENTS   
2354        GOOD             40             1212  SINGLE FAM RES IMPROVEMENTS   

      construction_type_code  ...                              bsmt_type  \
2839                   310.0  ...                                    NaN   
3656                   310.0  ...  SUBTERRANEAN BASEMENT UNFINISHED AREA   
3310                   310.0  ...                                    NaN   
1824                   310.0  ...        WALK-OUT BASEMENT FINISHED AREA   
2354                   310.0  ...                                    NaN   

      car_storage_type_code  car_storage_type  ext_wall_code  \
2839                    NaN               NaN           10.0   
3656                    GRD   DETACHED GARAGE           10.0   
3310                    GRC      CARPORT AREA          100.0   
1824                    GRD   DETACHED GARAGE           30.0   
2354                    GRD   DETACHED GARAGE           30.0   

                 ext_wall     ac  heating  bsmt_finished  num_buildings  \
2839     FRAME WOOD/SHAKE  False     True          False              2   
3656     FRAME WOOD/SHAKE  False     True          False              2   
3310  MOSS ROCK/FLAGSTONE  False     True          False              1   
1824         FRAME STUCCO  False     True           True              2   
2354         FRAME STUCCO   True     True          False              1   

     num_bath  
2839     1.00  
3656     1.00  
3310     1.50  
1824     2.25  
2354     5.25  

[5 rows x 30 columns]

In [178]:
building_dwellings = building_dwellings.loc[building_dwellings.groupby('strap')['bld_num'].idxmin()]
building_dwellings.groupby('strap').count()['bld_num'].value_counts()

# Note that all straps only have one occurrence now

1    3313
Name: bld_num, dtype: int64

Now that we have a data frame that has been preprocessed for our needs, we can join it with `sales_total` and `land`.

In [179]:
sales_total = sales_adj.merge(building_dwellings[['strap', 'total_finished_sqft', 'effective_year_built', 'quality', 'quality_code', 'num_bed', 'num_bath', 'bsmt_finished', 'num_buildings']],
                              on = 'strap',
                              how = 'left').merge(land[['strap', 'sqft']],
                                                  on = 'strap',
                                                  how = 'left')
sales_total.rename(columns = {'sqft': 'total_land_sqft'}, inplace = True)
sales_total

strap sales_cd sales_cd_dscr    price  market_area  multiplier  \
0     R0000008        Q     qualified    65000          102         NaN   
1     R0000019        U   unqualified    75000          109         NaN   
2     R0000019        U   unqualified   110600          109         NaN   
3     R0000019        Q     qualified   126400          109         NaN   
4     R0000019        Q     qualified   332000          109         NaN   
...        ...      ...           ...      ...          ...         ...   
9598  R0610553        Q     qualified  2872600          102      1.2167   
9599  R0610553        Q     qualified  3450000          102         NaN   
9600  R0612718        U   unqualified   910000          109      1.1910   
9601  R0613548        Q     qualified   850000          102      1.3453   
9602  R0616832        Q     qualified   710000          101         NaN   

      adjusted_sales_price  total_finished_sqft  effective_year_built  \
0                    65000               1558.0                1954.0   
1                    75000                942.0                1970.0   
2                   110600                942.0                1970.0   
3                   126400                942.0                1970.0   
4                   332000                942.0                1970.0   
...                    ...                  ...                   ...   
9598               3495092               2611.0                2021.0   
9599               3450000               2611.0                2021.0   
9600               1083810               2633.0                2024.0   
9601               1143505               1806.0                1988.0   
9602                710000                775.0                1980.0   

         quality  quality_code  num_bed  num_bath bsmt_finished  \
0     AVERAGE +           31.0      3.0      1.50         False   
1       AVERAGE           30.0      4.0      1.75          True   
2       AVERAGE           30.0      4.0      1.75          True   
3       AVERAGE           30.0      4.0      1.75          True   
4       AVERAGE           30.0      4.0      1.75          True   
...          ...           ...      ...       ...           ...   
9598    GOOD ++           42.0      4.0      5.25          True   
9599    GOOD ++           42.0      4.0      5.25          True   
9600  EXCELLENT           60.0      0.0      0.00          True   
9601  AVERAGE +           31.0      4.0      2.25         False   
9602    AVERAGE           30.0      2.0      1.00         False   

      num_buildings  total_land_sqft  
0               1.0             6801  
1               1.0             3606  
2               1.0             3606  
3               1.0             3606  
4               1.0             3606  
...             ...              ...  
9598            1.0             6987  
9599            1.0             6987  
9600            1.0             7730  
9601            1.0            10972  
9602            1.0             6236  

[9603 rows x 16 columns]

For the scope of this analysis, we are only interested in *qualified* sales, so we can apply a filter and save the resulting data frame for further analysis.

In [180]:
qualified_sales = sales_total[sales_total['sales_cd'] == 'Q']
qualified_sales

strap sales_cd sales_cd_dscr    price  market_area  multiplier  \
0     R0000008        Q     qualified    65000          102         NaN   
3     R0000019        Q     qualified   126400          109         NaN   
4     R0000019        Q     qualified   332000          109         NaN   
7     R0000021        Q     qualified   825000          105         NaN   
8     R0000021        Q     qualified   890000          105         NaN   
...        ...      ...           ...      ...          ...         ...   
9597  R0608224        Q     qualified  1050000          105      1.5029   
9598  R0610553        Q     qualified  2872600          102      1.2167   
9599  R0610553        Q     qualified  3450000          102         NaN   
9601  R0613548        Q     qualified   850000          102      1.3453   
9602  R0616832        Q     qualified   710000          101         NaN   

      adjusted_sales_price  total_finished_sqft  effective_year_built  \
0                    65000               1558.0                1954.0   
3                   126400                942.0                1970.0   
4                   332000                942.0                1970.0   
7                   825000               2042.0                1990.0   
8                   890000               2042.0                1990.0   
...                    ...                  ...                   ...   
9597               1578045               1796.0                2005.0   
9598               3495092               2611.0                2021.0   
9599               3450000               2611.0                2021.0   
9601               1143505               1806.0                1988.0   
9602                710000                775.0                1980.0   

         quality  quality_code  num_bed  num_bath bsmt_finished  \
0     AVERAGE +           31.0      3.0      1.50         False   
3       AVERAGE           30.0      4.0      1.75          True   
4       AVERAGE           30.0      4.0      1.75          True   
7        GOOD +           41.0      3.0      3.00          True   
8        GOOD +           41.0      3.0      3.00          True   
...          ...           ...      ...       ...           ...   
9597       GOOD           40.0      5.0      3.00         False   
9598    GOOD ++           42.0      4.0      5.25          True   
9599    GOOD ++           42.0      4.0      5.25          True   
9601  AVERAGE +           31.0      4.0      2.25         False   
9602    AVERAGE           30.0      2.0      1.00         False   

      num_buildings  total_land_sqft  
0               1.0             6801  
3               1.0             3606  
4               1.0             3606  
7               2.0             6884  
8               2.0             6884  
...             ...              ...  
9597            1.0             7445  
9598            1.0             6987  
9599            1.0             6987  
9601            1.0            10972  
9602            1.0             6236  

[5868 rows x 16 columns]

<span  style="color:purple; font-size:25px">
Exploring the Data
</span>

################## BREAK ########################### </br>
The data frame with all the relevent variables for plotting (what they recommended plus I added a num_buildings column) is in `qualified_sales`